### ISSUE TITLE
[Feature]: support true deletion of run / history

### ISSUE THREAD

**Issue by adefossez** [wandb/client#3424]
### Description

I am the author of Dora (https://github.com/facebookresearch/dora), a framework for experiment and grid search management. A key notion in Dora is that experiments are uniquely identified by a hash of their hyper parameters (more precisely, the hyper parameters that differ from a base config). This means for instance that when scheduling twice the same grid search, Dora can automatically find that a given set of hyper-parameters has already been scheduled, and avoid entirely duplicated runs issues. This hash is also a very convenient way of identifying an experiment, that will be constant across users and time.

I would like to use the Dora hash signature as an id in wandb, as this makes a number of things more convenient (like knowing easily the URL for a given experiment). However, it is currently not possible because there is no way to truely erase a run or its history from WanDB. In particular, if I run an experiment with a given Dora signature, and then wipe out its checkpoint (for instance there was a bug in the implementation), then I cannot reuse the same id with wandb. Setting `resume = False` won't solve the issue, as this doesn't wipe out the history, only seem to overwrite gradually which is not really nice.

Of course I could use a different WanDB id each time, but I thought in my case it it much nicer to have the possibility to use the Dora signature hash, which saves me the trouble of having to save anywhere the mapping with a WanDb id. I also tried deleting the run with `run.delete()` but this is not a true delete as it doesn't really free the id for later reuse.

### Suggested Solution

Give the freedom for your users to really delete entries. Give the freedom to your users to rewrite history as they see fit. I get it that you want to enforce some safety for the majority of the people, but for some specific use case and integration with other tools, allowing for more control over what is stored would be awesome.

**Comment by ramit-wandb** [comment]
Hi @adefossez,

Thank you for your suggestion! I have made a request for our engineering team to review this. I'll keep you updated on the status of this feature.

Thanks,
Ramit

**Comment by vanpelt** [comment]
Hey @adefossez something definitely on our roadmap.  Unfortunately it's very complicated because data is keyed off of this ID in many other datastores like the history of metrics and the files in object stores.  The ultimate solution will likely still have a delay to support "unarchiving".  Do you think it would be possible to add a single additional byte to the id to account for this like "version" or something equivalent that could be incremented?

**Comment by JacobARose** [comment]
> Hey @adefossez something definitely on our roadmap. Unfortunately it's very complicated because data is keyed off of this ID in many other datastores like the history of metrics and the files in object stores. The ultimate solution will likely still have a delay to support "unarchiving". Do you think it would be possible to add a single additional byte to the id to account for this like "version" or something equivalent that could be incremented?

I believe if wandb offered an intuitive way to automatically increment an experiment's version if it detects it's been run before, that would go a long way toward reducing the overall bloat/ambiguity that comes from an evolving code base being used to execute complex experiments. 

A likely related but orthogonal problem: I've frequently had to delete entire projects b/c of earlier runs that used slight variations of the name of a logged metric. Such as changing "train/F1_epoch" to "train_F1_epoch", both will take up space in the project table forever.

**Comment by adefossez** [comment]
@vanpelt thanks for looking into this. In my case, the extra bits would necessarily mean that I have to keep track somewhere of the version count.
However, most of my backend is just files on NFS, not a fancy dB, because it is the easiest to have on our research cluster, so keeping track of version number is not easy. I can't prevent the end user from deleting entire experiment folders, and later run the same experiment again.
Because the experiment id is a hash of the experiment hyper params, the exact same id would be reused. This experiment id to experiment mapping is thus completely stateless, and cannot be altered unless I make it state full, losing its simplicity and the fact that a central db is not required.

I could assign a second id just for Wandb but that makes it a lot less convenient. The advantage of the stateless experiment id computed from its params is that if my coworker run the same XP, they get the same id. Thus we can use this id to talk in a non ambiguous way about the same XP, even without a centralized db. Having an extra wandb id mapping means we can't guess the url on wandb without a centralized db, which means extra complexity.

I see that it is already possible to override metrics values of past epoch. Why not also store a "max epoch" field and allow only that to go back to 0 to truncate at least the plots ?

Im guessing you do not want to cascade delete to avoid people destroying their research history by mistake ?

**Comment by vanpelt** [comment]
@adefossez yeah it's complicated.  What about using wandb as the state storage system.  You could add something like "generations" to your config.  And then do something like:

```python
import wandb

api = wandb.Api()
RUN_ID = "experimentid"
generations = 1
try:
   run = api.run("ENTITY/PROJECT/RUN_ID")
   run.config["generations"] += 1
   generations = run.config["generations"]
   run.update()
   RUN_ID += str(run.config["generations"])
except wandb.CommError:
   pass
   
wandb.init(id=RUN_ID, config={"generations": generations})
```

**Comment by MBakirWB** [comment]
HI @adefossez , 

Following up this thread. Please let us know if you have any questions and/or require further assistance or feedback.

**Comment by adefossez** [comment]
Storing state in wandb is possible however it still breaks the predictability of wandb ids without calling first its API. 
I still think it is a strange choice not to have any support for deletion of the history, especially since the init API let us reuse an ID with "resume=false". Why provide this resume=false possibility if the support for it is broken ? If resume = false then clearly the metrics from a previous run are invalid and should not be shown anymore.

**Comment by adefossez** [comment]
@vanpelt  @MBakirWB  Pinging on this. I guess I made it too complex in a way the first time I described my issue.

`wandb.init` allows to give `resume=False` as an argument, to indicate not to resume the previous run. If a previous run existed, its metrics will still appear in the UI. The metrics will be slowly overwritten, one epoch by one epoch, as the new run progresses. This is quite confusing, and I would expect the metrics to be overwritten at once when passing `resume=False`. A solution in my opinion would be to store a `last_epoch` field in the dB. When one passes `resume=False` (or maybe once the first push of metrics happens) , the `last_epoch` resets to 0. In the UI, only show the points for the range `[0, last_epoch]`.

The current UI when using `resume=False` is confusing, showing a mixture of new and old metrics.

**Comment by vanpelt** [comment]
@adefossez understood, we're working on a specific feature for this usecase.  We're planning to add a method like `run.rewind(...)` that will allow you to got to a specific step and overwrite all data moving forward.

**Comment by adefossez** [comment]
That sounds perfect, looking forward to the release!

**Comment by PheelaV** [comment]
Hi, I wanted to ask what is the state of the `run.rewind(...)`? It still shows up as a "private preview" feature.

**Comment by haok1402** [comment]
For over three years since this issue was first opened, I can't believe we're still at a stats quo...

---

For future users, I ended up doing this. Not elegant; but allows me to re-use the run-id.

1. Move all the runs except the one to delete to a new project.
2. Delete the old project (i.e. so the run to delete is truely deleted).
3. Rename the new project back with the old name.
